tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/WeightsBroadcastOps.java

```diff
@@ -0,0 +1,186 @@
+                weightsRankStatic,
+                valuesShapeStatic.toString(),
+                weightsShapeStatic.toString()));
+      }
+
+      for (int i = 0; i < valuesRankStatic; i++) {
+        if (valuesShapeStatic.size(i) != weightsShapeStatic.size(i)) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "%s Mismatch at dim %s. values.shape=%s weights.shape=%s.",
+                  ASSERT_BROADCASTABLE_ERROR_PREFIX,
+                  i,
+                  valuesShapeStatic.toString(),
+                  weightsShapeStatic.toString()));
+        }
+      }
+      return tf.withSubScope("staticDimsCheckSuccess")
+          .withControlDependencies(Collections.EMPTY_LIST)
+          .noOp();
+    }
+    // Dynamic checks.
+    Operand<TBool> is_scalar = tf.math.equal(weightsRank, tf.constant(0));
+    List<Operand<?>> data =
+        Arrays.asList(
+            tf.constant(ASSERT_BROADCASTABLE_ERROR_PREFIX),
+            tf.constant("weights.shape="),
+            weightsShape,
+            tf.constant("values.shape="),
+            valuesShape,
+            tf.constant("is_scalar="),
+            is_scalar);
+
+    Operand<TBool> isValidShape =
+        tf.select(
+            is_scalar,
+            is_scalar,
+            hasValidNonscalarShape(tf, weightsRank, weightsShape, valuesRank, valuesShape));
+
+    return tf.assertThat(isValidShape, data);
+  }
+
+  /**
+   * Check to see that weights and values have the same rank, if they do, then check each
+   * corresponding dim of each.
+   *
+   * @param tf The TensorFlow Ops
+   * @param weightsRank the rank operand for the weights
+   * @param weightsShape the shape operand for the weights
+   * @param valuesRank the rank operand for the values
+   * @param valuesShape the shape operand for the values
+   * @return a boolean Operand, true if both shapes have the same rank, and each dimension is the
+   *     same
+   */
+  private static Operand<TBool> hasValidNonscalarShape(
+      Ops tf,
+      Operand<TInt32> weightsRank,
+      Operand<TInt32> weightsShape,
+      Operand<TInt32> valuesRank,
+      Operand<TInt32> valuesShape) {
+    tf = tf.withSubScope("hasValidNonscalarShape");
+    Operand<TBool> isSameRank = tf.math.equal(valuesRank, weightsRank);
+    return tf.select(isSameRank, hasValidDims(tf, weightsShape, valuesShape), isSameRank);
+  }
+
+  /**
+   * Checks that each dimension of the two shapes are the same
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsShape the shape of the weights
+   * @param valuesShape the shape of the values
+   * @return a boolean Operand, true if all the dimensions of the two shapes are the same.
+   */
+  private static Operand<TBool> hasValidDims(
+      Ops tf, Operand<TInt32> weightsShape, Operand<TInt32> valuesShape) {
+    tf = tf.withSubScope("hasInvalidDims");
+    Operand<TInt32> diff = tf.reduceSum(tf.math.sub(weightsShape, valuesShape), tf.constant(0));
```

Comment:
This seems to compute the sum of the element-wise differences in sizes, which is not at all what's needed.

Reply:
Let me rewrite, it looks like the python code is different. Not sure what happened here.